src/server/tiering/small_bins.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/small_bins.h"
+
+#include <algorithm>
+#include <optional>
+#include <utility>
+
+#include "absl/base/internal/endian.h"
+#include "base/logging.h"
+#include "core/compact_object.h"
+#include "server/tiering/disk_storage.h"
+
+namespace dfly::tiering {
+
+std::optional<SmallBins::FilledBin> SmallBins::Stash(std::string_view key, std::string_view value) {
+  DCHECK_LT(value.size(), 2_KB);
+
+  // See FlushBin() for format details
+  size_t value_bytes = 8 /* hash */ + 2 /* value length*/ + value.size();
+
+  std::optional<FilledBin> filled_bin;
+  if (2 /* num entries */ + current_bin_bytes_ + value_bytes >= 4_KB) {
+    filled_bin = FlushBin();
+  }
+
+  current_bin_bytes_ += value_bytes;
+  current_bin_[key] = value;
+  return filled_bin;
+}
+
+SmallBins::FilledBin SmallBins::FlushBin() {
+  std::string out;
+  out.resize(current_bin_bytes_ + 1);
+
+  unsigned id = ++last_bin_id_;
+  auto& pending_set = pending_bins_[id];
+
+  char* data = out.data();
+
+  // Store number of entries, 2 bytes
```

Comment:
What is the reason that you changed the format?
why do you need to store number of items and value length inside the page?

Reply:
Because we don't divide by value length into different bins. Let's please just make it work and optimize it later 🥲

Reply:
I missed your comment from yesterday. It's ok but it caught me by surprise since we have not discussed this